exec/cnex/lib/os_posix.c

```diff
@@ -0,0 +1,197 @@
+    }
+    return pp->ptr;
+}
+
+void object_releaseProcessObject(Object *o)
+{
+    if (o != NULL) {
+        assert(o->refcount > 0);
+        o->refcount--;
+
+        if (o->refcount <= 0) {
+            if (o->ptr != NULL) {
+                if (((TProcessHandle*)o->ptr)->proc != 0) {
+                    close(((TProcessHandle*)o->ptr)->proc);
+                }
+                free(o->ptr);
+            }
+            free(o);
+        }
+    }
+}
+
+Cell *object_processObjectToString(Object *self)
+{
+    Cell *r = cell_fromCString("<PROCESS ");
+    char s[16];
+    sprintf(s, "%d", (pid_t)((TProcessHandle*)self->ptr)->proc);
+    r->string = string_appendCString(r->string, s);
+    r->string = string_appendCString(r->string, ">");
+    return r;
+}
+
+Object *object_createProcessObject(void *p)
+{
+    Object *r = object_createObject();
+    r->ptr = p;
+
+    r->release = object_releaseProcessObject;
+    r->toString = object_processObjectToString;
+    return r;
+}
+
+void closer()
+{
+    if (g_children == NULL  || g_children->data == NULL) {
+        return;
+    }
+    for (size_t i = 0; i < g_children->size; i++) {
+        if (g_children->data[i].type != cNothing) {
+            TProcessHandle *p = g_children->data[i].object->ptr;
+            kill(-(p->proc), 9);
+        }
+    }
+}
+
+
+
+void os_initModule()
+{
+    g_children = array_createArray();
+}
+
+void os_shutdownModule()
+{
+    array_freeArray(g_children);
+    g_children = NULL;
+}
+
+
+void os_chdir(TExecutor *exec)
+{
+    char *path = string_asCString(top(exec->stack)->string); pop(exec->stack);
+
+    int r = chdir(path);
+    if (r != 0) {
+        exec->rtl_raise(exec, "OsException.PathNotFound", path, BID_ZERO);
+    }
+    free(path);
+}
+
+void os_getcwd(TExecutor *exec)
+{
+    char buf[MAXPATHLEN];
+    getcwd(buf, sizeof(buf));
+
+    push(exec->stack, cell_fromString(string_createCString(buf)));
+}
+
+void os_kill(TExecutor *exec)
+{
+    Object *process = top(exec->stack)->object; pop(exec->stack);
+    TProcessHandle *p = check_process(exec, process);
+
+    kill(-(pid_t)p->proc, SIGTERM);
+    p->proc = 0;
+    for (size_t i = 0; i < g_children->size; i++) {
+        if (g_children->data[i].type == cObject) {
+            Object *pi = g_children->data[i].object;
+            if (pi->ptr == p) {
+                cell_clearCell(&g_children->data[i]);
+                break;
+            }
+        }
+    }
+}
+
+void os_platform(TExecutor *exec)
+{
+    push(exec->stack, cell_fromNumber(number_from_uint32(ENUM_Platform_posix)));
+}
+
+void os_spawn(TExecutor *exec)
+{
+    char *cmd = string_asCString(top(exec->stack)->string); pop(exec->stack);
+
+    static BOOL init_closer = FALSE;
+    if (init_closer == FALSE) {
+        atexit(closer);
+        init_closer = TRUE;
+    }
+
+    TProcessHandle *pid = malloc(sizeof(TProcessHandle));
+    pid->proc = fork();
+    if (pid->proc == 0) {
+        setpgid(0, 0);
+        array_clearArray(g_children);
```

Comment:
There's no need to clear this array here. You're about to call exec, so any memory you might have allocated anywhere is completely irrelevant.

Reply:
What if you're calling exec twice in the same application?  Will it still be ok?  In fact.  Would it be better to eliminate the os_moduleInit(), routine, and just initialize the g_children array right at this point here?